controller/src/controller.rs

```diff
@@ -989,6 +994,35 @@ impl Controller {
     // Implementation of the read function, which does not check that the memory
     // pages addressed by `read` are valid for the addressed modules.
     async fn read_impl(&self, modules: ModuleId, read: MemoryRead) -> Result<ReadResult, Error> {
+        // It's possible for consumers to request more data than will fit in a
+        // single protocol message. We'll split that up here, rather than
+        // requiring the consumer to worry about it.
+        let mut split_modules = split_large_reads(&modules, &read).into_iter();
+        // Self::read() must guarantee that we have at least one module and so
+        // at least one read.
+        let Some(first_modules) = split_modules.next() else {
+            panic!("Split into zero modules? modules {modules:?}, read {read:?}");
+        };
+        let mut response = self.read_impl_one_message(first_modules, read).await?;
+        for next_modules in split_modules {
+            let next_response = self.read_impl_one_message(next_modules, read).await?;
+            response = response.merge(&next_response).unwrap();
+        }
+        Ok(response)
```

Comment:
I spent a while bashing my head against `async` to see if we could do this without explicit iteration.

The answer is "yes", although I'm not convinced that it's worthwhile:
```rust
        // It's possible for consumers to request more data than will fit in a
        // single protocol message. We'll split that up here, rather than
        // requiring the consumer to worry about it.
        //
        // Self::read() must guarantee that we have at least one module and so
        // at least one read, so we can unwrap the result.
        use futures::StreamExt;
        use stream_reduce::Reduce;
        futures::stream::iter(split_large_reads(&modules, &read))
            .then(|m| self.read_impl_one_message(m, read))
            .reduce(|a, b| async { Ok(a?.merge(&b?).unwrap()) })
            .await
            .expect("Split into zero modules? modules {modules:?}, read {read:?}")
```

(it requires adding `futures = "0.3"` and `stream-reduce = "0.1.0"` to the dependencies)

Reply:
Agreed about not being convinced, that's pretty terse :)

I'll stare at it a bit more, but maybe stick with the simpler / dumber case.